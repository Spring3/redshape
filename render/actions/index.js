import issueActions from './issue.actions';
import timeEntryActions from './timeEntry.actions';
import settingsActions from './settings.actions';

export default {
  issue: issueActions,
  timeEntry: timeEntryActions,
  settings: settingsActions
};
