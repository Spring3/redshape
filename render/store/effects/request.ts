import { ipcRenderer, remote } from 'electron';
import { Response } from '../../../types';

const crypto = remote.require('crypto');

type Config = {
  endpoint: string;
  token?: string;
}

type QueryConfig = {
  payload: Record<string, any>;
  config?: Config;
}

const query = (config: QueryConfig): Promise<Response> => {
  const id = crypto.randomBytes(10).toString('hex');

  return new Promise((resolve) => {
    ipcRenderer.send('request', JSON.stringify({ ...config, id }));

    ipcRenderer.once(`response:${id}`, (event, response) => {
      console.log('response:', response);
      resolve(response);
    });
  });
};

export {
  query,
};

export type {
  QueryConfig
};
