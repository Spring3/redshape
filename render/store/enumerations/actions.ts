import type { IAction, Context } from 'overmind';
import { Activity, Response } from '../../../types';

const fetchActivities: IAction<void, Promise<Response<Activity[]>>> = async ({ effects, state }: Context) => {
  const response = await effects.mainProcess.request({
    payload: {
      method: 'GET',
      route: 'enumerations/time_entry_activities.json',
    }
  });

  if (response.success) {
    state.enumerations.activities = response.data.activities as Activity[] ?? [];
    return {
      success: true,
      data: state.enumerations.activities,
    };
  }

  return {
    success: false,
    error: response.error
  };
};

export {
  fetchActivities
};
