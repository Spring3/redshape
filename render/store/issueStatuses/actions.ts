import { IAction, Context } from 'overmind';
import { IssueStatus, PaginatedActionResponse } from '../../../types';
import { indexById } from '../helpers';

const getAll: IAction<void, Promise<PaginatedActionResponse<IssueStatus>>> = async ({ effects, state }: Context) => {
  const response = await effects.mainProcess.request({
    payload: {
      method: 'GET',
      route: 'issue_statuses.json'
    }
  });

  if (response.success) {
    state.issueStatuses.byId = {
      ...state.issueStatuses,
      ...indexById(response.payload.issueStatuses)
    };

    return {
      success: true,
      data: {
        items: response.payload.issueStatuses,
        total: response.payload.total,
        limit: response.payload.limit,
        offset: response.payload.offset
      },
      hasMore: response.payload.total > (response.payload.offset + response.payload.issueStatuses.length)
    };
  }

  return {
    success: false,
    data: {
      items: [],
      total: 0,
      limit: response.payload.limit,
      offset: response.payload.offset
    },
    hasMore: false,
    error: response.error
  };
};

export {
  getAll
};
