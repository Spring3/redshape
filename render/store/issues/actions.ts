import type { Context, IAction } from 'overmind';
import { Issue, PaginatedActionResponse, Response } from '../../../types';
import { indexById } from '../helpers';

type GetManyIssuesArgs = {
  filters: {
    assigned_to_id?: number | string;
    status_id?: '*' | 'open' | 'closed';
    subject?: string;
    sort?: string;
  },
  offset?: number;
  limit?: number;
}

const getMany: IAction<GetManyIssuesArgs, Promise<PaginatedActionResponse<Issue>>> = async ({ effects, state }: Context, { filters, offset = 0, limit = 20 }) => {
  const query = {
    limit: limit ? Math.abs(limit) : undefined,
    offset: offset ? Math.abs(offset) : 0,
    include: 'attachments,relations',
    ...filters
  };

  const response = await effects.mainProcess.request({
    payload: {
      route: 'issues.json',
      method: 'GET',
      query
    }
  });

  if (response.success) {
    state.issues.byId = {
      ...state.issues.byId,
      ...indexById(response.payload.issues)
    };

    console.log('payload', response.payload);

    return {
      success: true,
      data: {
        items: response.payload.issues,
        total: response.payload.total,
        limit: response.payload.limit,
        offset: response.payload.offset
      },
      hasMore: response.payload.total > (response.payload.offset + response.payload.issues.length),
    };
  }

  return {
    success: false,
    data: {
      items: [],
      total: 0,
      limit,
      offset
    },
    hasMore: false,
    error: response.error
  };
};

type GetOneIssueArgs = {
  id: string;
}

const getOne: IAction<GetOneIssueArgs, Promise<Response<Issue>>> = async ({ effects, state }: Context, { id }) => {
  const response = await effects.mainProcess.request({
    payload: {
      route: `issues/${id}.json`,
      method: 'GET',
      query: {
        include: 'attachments,children,relations,journals',
      }
    }
  });

  if (response.success) {
    const issue = response.payload as Issue;
    state.issues.byId[id] = issue;
    return {
      success: true,
      data: issue
    };
  }

  return {
    success: false,
    error: response.error
  };
};

export {
  getOne,
  getMany
};
