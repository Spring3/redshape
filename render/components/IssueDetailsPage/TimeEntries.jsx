import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { connect } from 'react-redux';
import styled, { css, withTheme } from 'styled-components';

import PlusIcon from 'mdi-react/PlusIcon';
import CloseIcon from 'mdi-react/CloseIcon';
import TimerIcon from 'mdi-react/TimerIcon';

import InfiniteScroll from '../InfiniteScroll';
import ProcessIndicator, { OverlayProcessIndicator } from '../ProcessIndicator';
import Button, { GhostButton } from '../Button';
import DateComponent from '../Date';
import Dialog from '../Dialog';
import actions from '../../actions';

const HeaderContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  border-radius: 3px;

  h2 {
    display: inline-block;
    margin-left: 15px;
    margin-right: 15px;
  }

  & > div {
    margin-right: 15px;
    padding: 0px 5px;
    border-radius: 3px;

    button {
      background: ${props => props.theme.bg};
      margin: 0px 5px;
    }
  }
`;

const FlexButton = styled(Button) `
  display: inline-flex;
  align-items: center;
  ${props => css`
    background: ${props.theme.mainLight};
  `}
`;

const TimeEntriesContainer = styled.div`
  background: white;
  padding-top: 35px;
  max-height: 550px;
  min-width: 350px;
`;

const TimeEntriesList = styled.ul`
  position: relative;
  list-style-type: none;
  padding: 0px 10px 10px 10px;
  border-radius: 3px;
  margin: 0;
  overflow-y: scroll;
  max-height: 500px;
  background: ${props => props.theme.bgDark};
  box-shadow: inset 0px 0px 10px 0px ${props => props.theme.bgDark};

  li {
    cursor: pointer;
    display: block;
    padding: 10px;
    margin: 10px auto 0px auto;
    border-radius: 3px;
    border: 2px solid transparent;
    background: ${props => props.theme.bg};

    div:first-child {
      display: flex;
      align-items: center;
      justify-content: space-between;

      a {
        margin-left: 10px;
        visibility: hidden;
      }

      div {
        flex-grow: 1;
        display: flex;
        justify-content: space-between;
        span.time,
        span.date {
          margin-right: 5px;
          font-size: 12px;
          color: ${props => props.theme.minorText};
        }

        span.username {
          font-weight: bold;
          margin-right: 5px;
          color: ${props => props.theme.normalText};
        }
      }
    }

    p {
      margin-bottom: 0px;
      min-width: 100%;
      width: 0;
      line-height: 2;
    }

    &:hover {
      div:first-child {
        a {
          visibility: visible;
        }
      }
    }
  }
`;

const ProcessIndicatorWrapper = styled.li`
  position: relative;
  div {
    position: absolute;
    left: 24%;
    bottom: 0;

    span {
      position: relative;
      bottom: 5px;
      left: 60px;
    }
  }
`;

class TimeEntries extends Component {
  constructor(props) {
    super(props);
    this.listRef = React.createRef();
  }
  componentWillMount() {
    const { fetchIssueTimeEntries, selectedIssue } = this.props;
    fetchIssueTimeEntries(selectedIssue.id, 0);
  }

  componentDidUpdate(oldProps) {
    if (oldProps.selectedIssue.id !== this.props.selectedIssue.id) {
      this.props.fetchIssueTimeEntries(this.props.selectedIssue.id, 0);
    }
  }

  openModal = timeEntry => () => {
    this.props.showTimeEntryModal(timeEntry);
  }

  startTimeTracking = () => {
    const { selectedIssue, startTimeTracking } = this.props;
    startTimeTracking(selectedIssue);
  }

  removeTimeEntry = timeEntryId => (e) => {
    e.preventDefault();
    e.stopPropagation();
    const { selectedIssue, removeTimeEntry } = this.props;
    removeTimeEntry(timeEntryId, selectedIssue.id);
  }

  loadSpentTime = () => {
    const { spentTime, selectedIssue } = this.props;
    const { page } = spentTime;
    this.props.fetchIssueTimeEntries(selectedIssue.id, page + 1);
  }

  render() {
    const { spentTime, userId, theme, isTimerEnabled, trackedIssueId, selectedIssue } = this.props;
    return (
      <TimeEntriesContainer>
        <HeaderContainer>
          <h2 className="padded">Time spent</h2>
          <div>
            <FlexButton onClick={this.openModal()}>
              <PlusIcon size={22} />
              <span>&nbsp;Add</span>
            </FlexButton>
            <FlexButton
              disabled={isTimerEnabled || trackedIssueId === selectedIssue.id}
              onClick={this.startTimeTracking}>
              <TimerIcon size={22} />
              <span>&nbsp;Track</span>
            </FlexButton>
          </div>
        </HeaderContainer>
        <TimeEntriesList ref={this.listRef}>
          <InfiniteScroll
            load={this.loadSpentTime}
            isEnd={spentTime.data.length === spentTime.totalCount}
            hasMore={!spentTime.isFetching && !spentTime.error && spentTime.data.length < spentTime.totalCount}
            loadIndicator={<ProcessIndicatorWrapper><ProcessIndicator /></ProcessIndicatorWrapper>}
            container={this.listRef.current}
            immediate={true}
          >
            {spentTime.data.map(timeEntry => (
              <li key={timeEntry.id} onClick={this.openModal(timeEntry)}>
                <div>
                  <div>
                    <span className="username">{timeEntry.user.name}</span>
                    <span className="time">{timeEntry.hours} hours</span>
                    <DateComponent className="date" date={timeEntry.spent_on} />
                  </div>
                  {
                    userId === timeEntry.user.id && (
                      <Dialog title="Please Confirm" message="Are you sure you want to delete this time entry?">
                        {
                          requestConfirmation => (
                            <GhostButton
                              onClick={requestConfirmation(this.removeTimeEntry(timeEntry.id))}
                            >
                              <CloseIcon color={theme.normalText} />
                            </GhostButton>
                          )
                        }
                      </Dialog>
                    )
                  }
                </div>
                <p>{timeEntry.comments}</p>
              </li>
            ))}
            </InfiniteScroll>
        </TimeEntriesList>
      </TimeEntriesContainer>
    );
  }
}

TimeEntries.propTypes = {
  selectedIssue: PropTypes.shape({
    id: PropTypes.number.isRequired,
    subject: PropTypes.string.isRequired,
    journals: PropTypes.arrayOf(PropTypes.object).isRequired,
    project: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    author: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  userId: PropTypes.number.isRequired,
  userName: PropTypes.string.isRequired,
  spentTime: PropTypes.arrayOf(PropTypes.shape({
    activity: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    comments: PropTypes.string,
    created_on: PropTypes.string.isRequired,
    hours: PropTypes.number.isRequired,
    id: PropTypes.number.isRequired,
    issue: PropTypes.shape({
      id: PropTypes.number.isRequired
    }).isRequired,
    project: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    spent_on: PropTypes.string.isRequired,
    user: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired
  })).isRequired,
  isTimerEnabled: PropTypes.bool.isRequired,
  trackedIssueId: PropTypes.number,
  removeTimeEntry: PropTypes.func.isRequired,
  fetchIssueTimeEntries: PropTypes.func.isRequired,
  showTimeEntryModal: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  userId: state.user.id,
  userName: state.user.name,
  spentTime: state.issues.selected.spentTime,
  selectedIssue: state.issues.selected.data,
  isTimerEnabled: state.tracking.isEnabled,
  trackedIssueId: state.tracking.issue.id
});

const mapDispatchToProps = dispatch => ({
  fetchIssueTimeEntries: (issueId, page) => dispatch(actions.issues.getTimeEntriesPage(issueId, undefined, page)),
  startTimeTracking: selectedIssue => dispatch(actions.tracking.trackingStart(selectedIssue)),
  removeTimeEntry: (timeEntryId, issueId) => dispatch(actions.timeEntry.remove(timeEntryId, issueId))
});

export default withTheme(connect(mapStateToProps, mapDispatchToProps)(TimeEntries));
