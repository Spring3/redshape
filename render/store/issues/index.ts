import { state, IssuesState } from './state';
import * as actions from './actions';

export {
  state,
  actions
};

export type {
  IssuesState
};
