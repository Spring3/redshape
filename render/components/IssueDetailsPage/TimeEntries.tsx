import React, { useCallback, useEffect, useRef } from 'react';
import { css as emotionCss } from '@emotion/react';
import styled, { css, useTheme } from 'styled-components';

import PlusIcon from 'mdi-react/PlusIcon';
import TimerIcon from 'mdi-react/TimerIcon';

import CloseIcon from 'mdi-react/CloseIcon';
import Button from '../Button';
import DateComponent from '../Date';
import { useOvermindActions, useOvermindState } from '../../store';
import { usePaginatedFetch } from '../../hooks/usePaginatedFetch';
import { theme as Theme } from '../../theme';
import { TimeEntry, TimeTrackingAction } from '../../../types';
import Dialog from '../Dialog';
import { GhostButton } from '../GhostButton';

const HeaderContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  border-radius: 3px;

  h2 {
    display: inline-block;
    margin-left: 15px;
    margin-right: 15px;
  }

  & > div {
    margin-right: 15px;
    padding: 0px 5px;
    border-radius: 3px;

    button {
      background: ${props => props.theme.bg};
      margin: 0px 5px;
    }
  }
`;

const FlexButton = styled(Button)`
  display: inline-flex;
  align-items: center;
  ${props => css`
    background: ${props.theme.mainLight};
  `}
`;

type TimeEntriesProps = {
  issueId: number;
};

const TimeEntries = ({ issueId }: TimeEntriesProps) => {
  const listRef = useRef<HTMLUListElement>(null);

  const state = useOvermindState();
  const actions = useOvermindActions();
  const theme = useTheme() as typeof Theme;

  const selectedIssue = state.issues.byId[issueId];
  const [lastRecord] = state.timeTracking.records.slice(-1);
  const isTimerEnabled = lastRecord?.action !== TimeTrackingAction.STOP;
  const trackedIssueId = lastRecord?.issueId;

  const requestTimeEntries = useCallback(
    params => actions.timeEntries.getManyTimeEntries({
      filters: {
        issueId: selectedIssue.id,
        projectId: selectedIssue.project.id,
        ...params.filters
      },
      limit: params.limit,
      offset: params.offset
    }),
    [selectedIssue]
  );

  const { items: timeEntries } = usePaginatedFetch<TimeEntry>({
    containerRef: listRef,
    request: requestTimeEntries
  });

  // openModal = (timeEntry) => () => {
  //   const { showTimeEntryModal } = this.props;
  //   showTimeEntryModal(timeEntry);
  // }

  const startTimeTracking = useCallback(() => {
    actions.timeTracking.track({ issueId });
  }, [issueId]);

  return (
    <div css={emotionCss`
      background: white;
      padding-top: 35px;
      max-height: 550px;
      min-width: 350px;
    `}
    >
      <HeaderContainer>
        <h2 className="padded">Time spent</h2>
        <div>
          <FlexButton
            id="openModal"
            onClick={() => {
              /* openModal */
            }}
          >
            <PlusIcon size={22} />
            <span>&nbsp;Add</span>
          </FlexButton>
          <FlexButton
            id="track"
            disabled={isTimerEnabled || trackedIssueId === selectedIssue.id}
            onClick={startTimeTracking}
          >
            <TimerIcon size={22} />
            <span>&nbsp;Track</span>
          </FlexButton>
        </div>
      </HeaderContainer>
      <ul
        css={emotionCss`
        position: relative;
        list-style-type: none;
        padding: 0px 10px 10px 10px;
        border-radius: 3px;
        margin: 0;
        overflow-y: scroll;
        max-height: 500px;
        background: ${theme.bgDark};
        box-shadow: inset 0px 0px 10px 0px ${theme.bgDark};
      `}
        ref={listRef}
        data-testId="time-entries"
      >
        {timeEntries.map(timeEntry => (
          // eslint-disable-next-line
          <li
            css={emotionCss`
              cursor: pointer;
              display: block;
              padding: 10px;
              margin: 10px auto 0px auto;
              border-radius: 3px;
              border: 1px solid ${theme.bgDarker};
              background: ${theme.bg};

              div:first-child {
                display: flex;
                align-items: center;
                justify-content: space-between;
          
                a {
                  margin-left: 10px;
                  visibility: hidden;
                }
          
                div {
                  flex-grow: 1;
                  display: flex;
                  justify-content: space-between;
                  span.time,
                  span.date {
                    margin-right: 5px;
                    font-size: 12px;
                    color: ${theme.minorText};
                  }
          
                  span.username {
                    font-weight: bold;
                    margin-right: 5px;
                    color: ${theme.normalText};
                  }
                }
              }

              &:hover {
                div:first-child {
                  a {
                    visibility: visible;
                  }
                }
              }
            `}
            key={timeEntry.id}
            onClick={() => {
              /* openModal(timeEntry) */
            }}
            data-testId="time-entry"
          >
            <div>
              <div>
                <span className="username">{timeEntry.user.name}</span>
                <span className="time">
                  {timeEntry.hours}
                  {' '}
                  hours
                </span>
                <DateComponent className="date" date={timeEntry.spentOn} />
              </div>
              {state.users.currentUser?.id === timeEntry.user.id && (
                <Dialog
                  onConfirm={() => actions.timeEntries.removeTimeEntry({ id: timeEntry.id })}
                  title="Please Confirm"
                  message="Are you sure you want to delete this time entry?"
                >
                  {(requestConfirmation: any) => (
                    <GhostButton id="confirmDeletion" onClick={requestConfirmation}>
                      <CloseIcon color={theme.normalText} />
                    </GhostButton>
                  )}
                </Dialog>
              )}
            </div>
            <p css={emotionCss`
              margin-bottom: 0px;
              min-width: 100%;
              width: 0;
              line-height: 2;
            `}
            >
              {timeEntry.comments}

            </p>
          </li>
        ))}
      </ul>
    </div>
  );
};
export { TimeEntries };
