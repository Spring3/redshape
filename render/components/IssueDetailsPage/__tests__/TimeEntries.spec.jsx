import React from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import MockAdapter from 'axios-mock-adapter';

import { TIME_ENTRY_DELETE } from '../../../actions/timeEntry.actions';
import { ISSUES_TIME_ENTRY_GET } from '../../../actions/issues.actions';
import { getInstance, reset, initialize } from '../../../../common/request';
import { theme } from '../../../theme';
import { TimeEntries } from '../TimeEntries';

const redmineEndpoint = 'redmine.test.test';
const token = 'multipass';
const mockStore = configureStore([thunk]);

let axiosMock;

describe('IssueDetails => TimeEntries componnet', () => {
  beforeAll(() => {
    initialize(redmineEndpoint, token);
    axiosMock = new MockAdapter(getInstance());
  });

  afterEach(() => {
    axiosMock.reset();
    cleanup();
  });

  afterAll(() => {
    axiosMock.restore();
    reset();
  });

  it('should fetch the time entries for the selected issue on mount', () => {
    axiosMock.onGet('/time_entries.json').reply(() => Promise.resolve([200]));
    const state = {
      user: {
        id: 1,
        name: 'John Wayne'
      },
      issues: {
        selected: {
          data: {
            id: 1,
            subject: 'Test'
          },
          spentTime: {
            data: [
              {
                id: 1,
                user: {
                  id: 1,
                  name: 'John Wayne'
                },
                activity: {
                  id: 1,
                  name: 'Development'
                },
                hours: 3.2,
                spent_on: '2011-01-01',
                comments: 'Let\'s get this done'
              }
            ]
          }
        }
      },
      tracking: {
        isEnabled: false,
        issue: {
          id: 1
        }
      }
    };
    const store = mockStore(state);
    render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <TimeEntries />
        </ThemeProvider>
      </Provider>
    );
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0].type).toBe(ISSUES_TIME_ENTRY_GET);
  });

  it('should be able to request the time entry modal', () => {
    axiosMock.onGet('/time_entries.json').reply(() => Promise.resolve([200]));
    const state = {
      user: {
        id: 1,
        name: 'John Wayne'
      },
      issues: {
        selected: {
          data: {
            id: 1,
            subject: 'Test'
          },
          spentTime: {
            data: [
              {
                id: 1,
                user: {
                  id: 1,
                  name: 'John Wayne'
                },
                activity: {
                  id: 1,
                  name: 'Development'
                },
                hours: 3.2,
                spent_on: '2011-01-01',
                comments: 'Let\'s get this done'
              }
            ]
          }
        }
      },
      tracking: {
        isEnabled: false,
        issue: {
          id: 1
        }
      }
    };
    const store = mockStore(state);
    const showTimeEntry = jest.fn();
    render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <TimeEntries showTimeEntryModal={showTimeEntry} />
        </ThemeProvider>
      </Provider>
    );
    fireEvent.click(document.querySelector('#openModal'));
    expect(showTimeEntry).toHaveBeenCalled();
  });

  it('should start the tracking on button click', () => {
    axiosMock.onGet('/time_entries.json').reply(() => Promise.resolve([200]));
    const state = {
      user: {
        id: 1,
        name: 'John Wayne'
      },
      issues: {
        selected: {
          data: {
            id: 1,
            subject: 'Test'
          },
          spentTime: {
            data: [
              {
                id: 1,
                user: {
                  id: 1,
                  name: 'John Wayne'
                },
                activity: {
                  id: 1,
                  name: 'Development'
                },
                hours: 3.2,
                spent_on: '2011-01-01',
                comments: 'Let\'s get this done'
              }
            ]
          }
        }
      },
      tracking: {
        isEnabled: false,
        issue: {
          id: 10
        }
      }
    };
    const store = mockStore(state);
    render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <TimeEntries />
        </ThemeProvider>
      </Provider>
    );
    fireEvent.click(document.querySelector('#track'));
  });

  it('should remove the time entry if the button is clicked', () => {
    axiosMock.onGet('/time_entries.json').reply(() => Promise.resolve([200]));
    axiosMock.onDelete('/time_entries/1.json').reply(() => Promise.resolve([200]));
    const state = {
      user: {
        id: 1,
        name: 'John Wayne'
      },
      issues: {
        selected: {
          data: {
            id: 1,
            subject: 'Test'
          },
          spentTime: {
            data: [
              {
                id: 1,
                user: {
                  id: 1,
                  name: 'John Wayne'
                },
                activity: {
                  id: 1,
                  name: 'Development'
                },
                hours: 3.2,
                spent_on: '2011-01-01',
                comments: 'Let\'s get this done'
              }
            ]
          }
        }
      },
      tracking: {
        isEnabled: false,
        issue: {
          id: 10
        }
      }
    };
    const store = mockStore(state);
    const showTimeEntryModal = jest.fn();
    render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <TimeEntries showTimeEntryModal={showTimeEntryModal} />
        </ThemeProvider>
      </Provider>
    );
    fireEvent.click(document.querySelector('#confirmDeletion'), {
      preventDefault: () => { /* noop */ },
      stopPropagation: () => { /* noop */ }
    });
    fireEvent.click(document.querySelector('#dialog-confirm'));
    expect(store.getActions().pop().type).toBe(TIME_ENTRY_DELETE);
  });
});
