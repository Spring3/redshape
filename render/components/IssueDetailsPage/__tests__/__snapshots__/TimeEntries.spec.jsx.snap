// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueDetails => TimeEntries component should match the snapshot 1`] = `
.c3 {
  border-radius: 3px;
  font-weight: bold;
  font-size: 14px;
  outline: none;
  text-align: center;
  background: #FFFFFF;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
  -webkit-transition: background ease .2s;
  transition: background ease .2s;
  width: auto;
  border: 2px solid #FF7079;
  color: #FF7079;
  cursor: pointer;
}

.c3:hover,
.c3:focus {
  background: #FF7079;
  color: #FFFFFF !important;
}

.c3:hover svg,
.c3:focus svg {
  fill: #FFFFFF;
}

.c3:active {
  background: #EF6069;
}

.c4 {
  border-radius: 3px;
  font-weight: bold;
  font-size: 14px;
  outline: none;
  text-align: center;
  background: #FFFFFF;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
  -webkit-transition: background ease .2s;
  transition: background ease .2s;
  width: auto;
  border: 2px solid #A0A0A0;
  background: #FFFFFF;
  color: #A0A0A0;
}

.c4 svg {
  fill: #A0A0A0;
}

.c7 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #FF7079;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
}

.c7:hover {
  color: #EF6069;
}

.c7:hover svg {
  fill: #EF6069;
}

.c7:active,
.c7:focus,
.c7:visited {
  background: transparent;
}

.c6 {
  position: relative;
  display: inline-grid;
  grid-template-rows: 1fr;
}

.c6 span {
  float: left;
  grid-row: 1;
  grid-column: 1;
}

.c6 span:last-child {
  opacity: 0;
  display: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border-radius: 3px;
}

.c1 h2 {
  display: inline-block;
  margin-left: 15px;
  margin-right: 15px;
}

.c1 > div {
  margin-right: 15px;
  padding: 0px 5px;
  border-radius: 3px;
}

.c1 > div button {
  background: #FFFFFF;
  margin: 0px 5px;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FF7F87;
}

.c0 {
  background: white;
  max-height: 550px;
  min-width: 350px;
}

.c5 {
  position: relative;
  list-style-type: none;
  padding: 0px 10px 10px 10px;
  border-radius: 3px;
  margin: 0;
  overflow-y: scroll;
  max-height: 500px;
  background: #F9F9F9;
  box-shadow: inset 0px 0px 10px 0px #F9F9F9;
}

.c5 li {
  cursor: pointer;
  display: block;
  padding: 10px;
  margin: 10px auto 0px auto;
  border-radius: 3px;
  border: 1px solid #E9E9E9;
  background: #FFFFFF;
}

.c5 li div.status {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c5 li div.status a {
  margin-left: 10px;
  visibility: hidden;
  position: absolute;
  right: 0;
  background: white;
  border-radius: 50%;
  border: 1px solid lightgray;
  margin-top: -16px;
}

.c5 li div.status div {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c5 li div.status div span.time,
.c5 li div.status div span.date {
  margin-right: 5px;
  font-size: 12px;
  color: #A0A0A0;
}

.c5 li div.status div span.username {
  font-weight: bold;
  margin-right: 5px;
  color: #3F3844;
  width: 150px;
}

.c5 li div.extra {
  display: grid;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  grid-template-columns: 1fr auto;
  grid-auto-flow: row;
  grid-column-gap: 1rem;
  grid-row-gap: 3px;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
}

.c5 li div.extra:not(:empty) {
  margin-top: 0.5rem;
}

.c5 li div.extra span {
  font-size: 12px;
  color: #A0A0A0;
}

.c5 li div.extra span:nth-child(even) {
  text-align: right;
}

.c5 li p {
  margin-bottom: 0px;
  min-width: 100%;
  width: 0;
  line-height: 2;
}

.c5 li:hover div:first-child a {
  visibility: visible;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <h2
      className="padded"
    >
      Time spent
    </h2>
    <div>
      <button
        className="c2 c3"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          className="mdi-icon "
          fill="currentColor"
          height={22}
          viewBox="0 0 24 24"
          width={22}
        >
          <path
            d="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
          />
        </svg>
        <span>
           Add
        </span>
      </button>
      <button
        className="c2 c4"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        <svg
          className="mdi-icon "
          fill="currentColor"
          height={22}
          viewBox="0 0 24 24"
          width={22}
        >
          <path
            d="M12,20A7,7 0 0,1 5,13A7,7 0 0,1 12,6A7,7 0 0,1 19,13A7,7 0 0,1 12,20M19.03,7.39L20.45,5.97C20,5.46 19.55,5 19.04,4.56L17.62,6C16.07,4.74 14.12,4 12,4A9,9 0 0,0 3,13A9,9 0 0,0 12,22C17,22 21,17.97 21,13C21,10.88 20.26,8.93 19.03,7.39M11,14H13V8H11M15,1H9V3H15V1Z"
          />
        </svg>
        <span>
           Track
        </span>
      </button>
    </div>
  </div>
  <ul
    className="c5"
  >
    <li
      onClick={[Function]}
    >
      <div
        className="status"
      >
        <div>
          <span
            className="username"
          >
            John Wayne
          </span>
          <span
            className="time"
          >
            3.2
             hours
          </span>
          <span
            className="c6"
          >
            <span
              align="right"
              className="date"
            >
              Jan 01 2011
            </span>
            <span
              className="date"
            >
              Jan 01 2011
            </span>
          </span>
        </div>
        <a
          className="c7"
          disabled={false}
          href="#"
          onClick={[Function]}
        >
          <svg
            className="mdi-icon "
            fill="#3F3844"
            height={24}
            viewBox="0 0 24 24"
            width={24}
          >
            <path
              d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
            />
          </svg>
        </a>
      </div>
      <p>
        Let's get this done
      </p>
      <div
        className="extra"
      >
        <span>
          Development
        </span>
      </div>
    </li>
    <li
      onClick={[Function]}
    >
      <div
        className="status"
      >
        <div>
          <span
            className="username"
          >
            John Snow
          </span>
          <span
            className="time"
          >
            3.2
             hours
          </span>
          <span
            className="c6"
          >
            <span
              align="right"
              className="date"
            >
              Jan 01 2011
            </span>
            <span
              className="date"
            >
              Jan 01 2011
            </span>
          </span>
        </div>
      </div>
      <p>
        I know nothing
      </p>
      <div
        className="extra"
      >
        <span>
          Testing
        </span>
      </div>
    </li>
  </ul>
</div>
`;
